experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts

```diff
@@ -0,0 +1,476 @@
+				const serializedFinalRanges: [LocalCompressedId, FinalCompressedId, FinalCompressedId][] = [];
+				for (const [alignedLocal, [firstFinal, lastFinal]] of entries(firstLocal, finalRangesTable)) {
+					serializedFinalRanges.push([alignedLocal, firstFinal, lastFinal]);
+				}
+				localRanges.push([firstLocal, lastLocal, serializedFinalRanges]);
+			} else {
+				localRanges.push([firstLocal, lastLocal]);
+			}
+		}
+		return serialized;
+	}
+
+	public static deserialize<TRangeObject>(
+		serialized: SerializedSessionIdNormalizer,
+		getRangeObject: (finalId: FinalCompressedId) => TRangeObject
+	): SessionIdNormalizer<TRangeObject> {
+		const normalizer = new SessionIdNormalizer<TRangeObject>();
+		const { idRanges } = normalizer;
+		for (const [firstLocal, lastLocal, serializedFinalRanges] of serialized.localRanges) {
+			let finalRanges: FinalRanges<TRangeObject> | undefined;
+			if (serializedFinalRanges !== undefined) {
+				assert(serializedFinalRanges.length !== 0, 'Empty range should not be reified.');
+				if (serializedFinalRanges.length === 1) {
+					const [_, firstFinal, lastFinal] = serializedFinalRanges[0];
+					finalRanges = [firstFinal, lastFinal, getRangeObject(firstFinal)];
+				} else {
+					finalRanges = SessionIdNormalizer.makeFinalRangesMap<TRangeObject>();
+					for (const [alignedLocal, firstFinal, lastFinal] of serializedFinalRanges) {
+						finalRanges.append(alignedLocal, [firstFinal, lastFinal, getRangeObject(firstFinal)]);
+					}
+				}
+			}
+			idRanges.append(firstLocal, [lastLocal, finalRanges]);
+		}
+		return normalizer;
+	}
+
+	public equals(
+		other: SessionIdNormalizer<TRangeObject>,
+		compareRangeObjects: (a: TRangeObject, b: TRangeObject) => boolean = () => true
+	): boolean {
+		return this.idRanges.equals(other.idRanges, (localRangeA, localRangeB) => {
+			const [lastLocalA, finalRangesA] = localRangeA;
+			const [lastLocalB, finalRangesB] = localRangeB;
+			if (finalRangesA === undefined || finalRangesB === undefined) {
+				return finalRangesA === finalRangesB;
+			}
+
+			const rangeEquals = (finalRangeA: FinalRange<TRangeObject>, finalRangeB: FinalRange<TRangeObject>) => {
+				const [firstFinalA, lastFinalA, rangeObjectA] = finalRangeA;
+				const [firstFinalB, lastFinalB, rangeObjectB] = finalRangeB;
+				return (
+					firstFinalA === firstFinalB &&
+					lastFinalA === lastFinalB &&
+					compareRangeObjects(rangeObjectA, rangeObjectB)
+				);
+			};
+
+			if (isSingleRange(finalRangesA) || isSingleRange(finalRangesB)) {
+				if (!isSingleRange(finalRangesA) || !isSingleRange(finalRangesB)) {
+					return false;
+				}
+				return rangeEquals(finalRangesA, finalRangesB);
+			}
+
+			return lastLocalA === lastLocalB && finalRangesA.equals(finalRangesB, rangeEquals);
+		});
+	}
+}
+
+/**
+ * Serialized table for normalizing IDs made by the local session.
+ *
+ * TODO: Move this into ID compressor persisted types when integrated.
+ */
+export interface SerializedSessionIdNormalizer {
```

Comment:
Doesn't need to be part of this PR, but it'd be cool to have a serialize and deserialize method on `AppendOnlySortedMap` itself. You could probably type it such that you only need to provide a `serializeKey` function if the key type is not a primitive, and only need to provide a `serializeValue` function if the value type is not a primitive.

Reply:
agreed, for another PR :)